File: android/app/tests/unit/src/com/android/bluetooth/bas/BatteryServiceTest.java
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.bas;

import static android.bluetooth.BluetoothDevice.BOND_BONDED;
import static android.bluetooth.BluetoothDevice.BOND_BONDING;
import static android.bluetooth.BluetoothDevice.BOND_NONE;
import static android.bluetooth.BluetoothProfile.CONNECTION_POLICY_ALLOWED;
import static android.bluetooth.BluetoothProfile.CONNECTION_POLICY_FORBIDDEN;
import static android.bluetooth.BluetoothProfile.CONNECTION_POLICY_UNKNOWN;
import static android.bluetooth.BluetoothProfile.STATE_DISCONNECTED;

import static com.android.bluetooth.TestUtils.MockitoRule;
import static com.android.bluetooth.TestUtils.getTestDevice;

import static com.google.common.truth.Truth.assertThat;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.doReturn;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothUuid;
import android.os.ParcelUuid;

import androidx.test.filters.MediumTest;

import com.android.bluetooth.TestLooper;
import com.android.bluetooth.TestUtils.MockitoRule;
import com.android.bluetooth.btservice.AdapterService;
import com.android.bluetooth.btservice.storage.DatabaseManager;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Mock;

import java.util.List;

@MediumTest
@RunWith(JUnit4.class)
public class BatteryServiceTest {
    @Rule public final MockitoRule mMockitoRule = new MockitoRule();

    @Mock private AdapterService mAdapterService;
    @Mock private DatabaseManager mDatabaseManager;

    private final BluetoothDevice mDevice = getTestDevice(78);

    private BatteryService mService;
    private TestLooper mLooper;

    @Before
    public void setUp() {
        mLooper = new TestLooper();

        doReturn(mDatabaseManager).when(mAdapterService).getDatabase();
        doReturn(BluetoothDevice.BOND_BONDED).when(mAdapterService).getBondState(any());

        mService = new BatteryService(mAdapterService, mLooper.getLooper());
        mService.setAvailable(true);
    }

    @After
    public void tearDown() {
        mService.cleanup();
        assertThat(BatteryService.getBatteryService()).isNull();
    }

    @Test
    public void getBatteryService() {
        assertThat(BatteryService.getBatteryService()).isEqualTo(mService);
    }

    @Test
    public void setConnectionPolicy() {
        assertThat(mService.setConnectionPolicy(mDevice, CONNECTION_POLICY_FORBIDDEN)).isTrue();
    }

    @Test
    public void getConnectionPolicy() {
        for (int policy :
                List.of(
                        CONNECTION_POLICY_UNKNOWN,
                        CONNECTION_POLICY_FORBIDDEN,
                        CONNECTION_POLICY_ALLOWED)) {
            doReturn(policy).when(mDatabaseManager).getProfileConnectionPolicy(any(), anyInt());
            assertThat(mService.getConnectionPolicy(mDevice)).isEqualTo(policy);
        }
    }

    @Test
    public void canConnect_whenNotBonded_returnFalse() {
        int badPolicyValue = 1024;
        int badBondState = 42;
        for (int bondState : List.of(BOND_NONE, BOND_BONDING, badBondState)) {
            for (int policy :
                    List.of(
                            CONNECTION_POLICY_UNKNOWN,
                            CONNECTION_POLICY_FORBIDDEN,
                            CONNECTION_POLICY_ALLOWED,
                            badPolicyValue)) {
                doReturn(bondState).when(mAdapterService).getBondState(any());
                doReturn(policy).when(mDatabaseManager).getProfileConnectionPolicy(any(), anyInt());
                assertThat(mService.canConnect(mDevice)).isEqualTo(false);
            }
        }
    }

    @Test
    public void canConnect_whenBonded() {
        int badPolicyValue = 1024;
        doReturn(BOND_BONDED).when(mAdapterService).getBondState(any());

        for (int policy : List.of(CONNECTION_POLICY_FORBIDDEN, badPolicyValue)) {
            doReturn(policy).when(mDatabaseManager).getProfileConnectionPolicy(any(), anyInt());
            assertThat(mService.canConnect(mDevice)).isEqualTo(false);
        }
        for (int policy : List.of(CONNECTION_POLICY_UNKNOWN, CONNECTION_POLICY_ALLOWED)) {
            doReturn(policy).when(mDatabaseManager).getProfileConnectionPolicy(any(), anyInt());
            assertThat(mService.canConnect(mDevice)).isEqualTo(true);
        }
    }

    @Test
    public void connectAndDump_doesNotCrash() {
        doReturn(CONNECTION_POLICY_ALLOWED)
                .when(mDatabaseManager)
                .getProfileConnectionPolicy(any(), anyInt());

        doReturn(new ParcelUuid[] {BluetoothUuid.BATTERY})
                .when(mAdapterService)
                .getRemoteUuids(any(BluetoothDevice.class));

        assertThat(mService.connect(mDevice)).isTrue();

        mService.dump(new StringBuilder());
    }

    @Test
    public void connect_whenForbiddenPolicy_FailsToConnect() {
        doReturn(CONNECTION_POLICY_FORBIDDEN)
                .when(mDatabaseManager)
                .getProfileConnectionPolicy(any(), anyInt());

        assertThat(mService.connect(mDevice)).isFalse();
    }

    @Test
    public void getConnectionState_whenNoDevicesAreConnected_returnsDisconnectedState() {
        assertThat(mService.getConnectionState(mDevice)).isEqualTo(STATE_DISCONNECTED);
    }

    @Test
    public void getDevices_whenNoDevicesAreConnected_returnsEmptyList() {
        assertThat(mService.getDevices()).isEmpty();
    }

    @Test
    public void getDevicesMatchingConnectionStates() {
        doReturn(new BluetoothDevice[] {mDevice}).when(mAdapterService).getBondedDevices();
        int[] states = new int[] {STATE_DISCONNECTED};

        assertThat(mService.getDevicesMatchingConnectionStates(states)).containsExactly(mDevice);
    }
}

File: android/app/tests/unit/src/com/android/bluetooth/bas/BatteryServiceTest.kt
/*
 * Copyright (C) 2026 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.bas

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothDevice.BOND_BONDED
import android.bluetooth.BluetoothDevice.BOND_BONDING
import android.bluetooth.BluetoothDevice.BOND_NONE
import android.bluetooth.BluetoothProfile.CONNECTION_POLICY_ALLOWED
import android.bluetooth.BluetoothProfile.CONNECTION_POLICY_FORBIDDEN
import android.bluetooth.BluetoothProfile.CONNECTION_POLICY_UNKNOWN
import android.bluetooth.BluetoothProfile.STATE_DISCONNECTED
import android.bluetooth.BluetoothUuid
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.MediumTest
import com.android.bluetooth.TestLooper
import com.android.bluetooth.btservice.AdapterService
import com.android.bluetooth.getTestDevice
import com.android.tests.bluetooth.MockitoRule
import com.google.common.truth.Truth.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.kotlin.any
import org.mockito.kotlin.whenever

/** Test cases for [BatteryService]. */
@MediumTest
@RunWith(AndroidJUnit4::class)
class BatteryServiceTest {
    @get:Rule val mockitoRule = MockitoRule()

    @Mock private lateinit var adapterService: AdapterService

    private val device = getTestDevice(78)

    private lateinit var looper: TestLooper
    private lateinit var service: BatteryService

    @Before
    fun setUp() {
        looper = TestLooper()

        doReturn(BOND_BONDED).whenever(adapterService).getBondState(any())

        service = BatteryService(adapterService, looper.looper)
        service.isAvailable = true
    }

    @After
    fun tearDown() {
        service.cleanup()
    }

    @Test
    fun setConnectionPolicy() {
        assertThat(service.setConnectionPolicy(device, CONNECTION_POLICY_FORBIDDEN)).isTrue()
    }

    @Test
    fun getConnectionPolicy() {
        listOf(CONNECTION_POLICY_UNKNOWN, CONNECTION_POLICY_FORBIDDEN, CONNECTION_POLICY_ALLOWED)
            .forEach { policy ->
                doReturn(policy)
                    .whenever(adapterService)
                    .getProfileConnectionPolicy(any(), any<Int>())
                assertThat(service.getConnectionPolicy(device)).isEqualTo(policy)
            }
    }

    @Test
    fun canConnect_whenNotBonded_returnFalse() {
        val badPolicyValue = 1024
        val badBondState = 42
        val policies =
            listOf(
                CONNECTION_POLICY_UNKNOWN,
                CONNECTION_POLICY_FORBIDDEN,
                CONNECTION_POLICY_ALLOWED,
                badPolicyValue,
            )
        listOf(BOND_NONE, BOND_BONDING, badBondState).forEach { bondState ->
            policies.forEach { policy ->
                doReturn(bondState).whenever(adapterService).getBondState(any())
                doReturn(policy)
                    .whenever(adapterService)
                    .getProfileConnectionPolicy(any(), any<Int>())
                assertThat(service.canConnect(device)).isFalse()
            }
        }
    }

    @Test
    fun canConnect_whenBonded() {
        val badPolicyValue = 1024
        doReturn(BOND_BONDED).whenever(adapterService).getBondState(any())

        listOf(CONNECTION_POLICY_FORBIDDEN, badPolicyValue).forEach { policy ->
            doReturn(policy).whenever(adapterService).getProfileConnectionPolicy(any(), any<Int>())
            assertThat(service.canConnect(device)).isFalse()
        }
        listOf(CONNECTION_POLICY_UNKNOWN, CONNECTION_POLICY_ALLOWED).forEach { policy ->
            doReturn(policy).whenever(adapterService).getProfileConnectionPolicy(any(), any<Int>())
            assertThat(service.canConnect(device)).isEqualTo(true)
        }
    }

    @Test
    fun connectAndDump_doesNotCrash() {
        doReturn(CONNECTION_POLICY_ALLOWED)
            .whenever(adapterService)
            .getProfileConnectionPolicy(any(), any<Int>())

        doReturn(arrayOf(BluetoothUuid.BATTERY))
            .whenever(adapterService)
            .getRemoteUuids(any<BluetoothDevice>())

        assertThat(service.connect(device)).isTrue()
        service.dump(StringBuilder())
    }

    @Test
    fun connect_whenForbiddenPolicy_FailsToConnect() {
        doReturn(CONNECTION_POLICY_FORBIDDEN)
            .whenever(adapterService)
            .getProfileConnectionPolicy(any(), any<Int>())

        assertThat(service.connect(device)).isFalse()
    }

    @Test
    fun getConnectionState_whenNoDevicesAreConnected_returnsDisconnectedState() {
        assertThat(service.getConnectionState(device)).isEqualTo(STATE_DISCONNECTED)
    }

    @Test
    fun getDevices_whenNoDevicesAreConnected_returnsEmptyList() {
        assertThat(service.getDevices()).isEmpty()
    }

    @Test
    fun getDevicesMatchingConnectionStates() {
        doReturn(setOf(device)).whenever(adapterService).bondedDevices
        val states = intArrayOf(STATE_DISCONNECTED)

        assertThat(service.getDevicesMatchingConnectionStates(states)).containsExactly(device)
    }
}
